### ISSUE TITLE
AR model with decimal array column is dirty right after load

### ISSUE THREAD

**Issue by johvet** [rails/rails#30487]
### Steps to reproduce

Here is a simple Rails 4.2.7 application demonstrating the problem. It contains failing specs already:
https://github.com/johvet/pg_array

For those not willing to mess with the above app, here's the problem.

* Create a model with an array column of decimals having a scale > 0
```
  class CreateSamples < ActiveRecord::Migration
    def change
      create_table :samples do |t|
        t.string  :name,  null: true
        t.decimal :value, precision: 18, scale: 6, array: true, default: [0], null: true
      end
    end
  end
```
* Create a record and inspect the changes
```
2.3.3 :001 > Sample.create(name: 'foo', value: [1.0, 2.3])
   (0.1ms)  BEGIN
  SQL (0.4ms)  INSERT INTO "samples" ("name", "value") VALUES ($1, $2) RETURNING "id"  [["name", "foo"], ["value", "{1.0,2.3}"]]
   (1.9ms)  COMMIT
 => #<Sample id: 2, name: "foo", value: [#<BigDecimal:7f8c7eb99418,'0.1E1',9(18)>, #<BigDecimal:7f8c7eb990f8,'0.23E1',18(27)>]> 
2.3.3 :002 > Sample.first
  Sample Load (0.6ms)  SELECT  "samples".* FROM "samples"  ORDER BY "samples"."id" ASC LIMIT 1
 => #<Sample id: 1, name: "foo", value: [#<BigDecimal:7f8c7eb2b990,'0.12E1',18(27)>, #<BigDecimal:7f8c7eb2b710,'0.3E1',9(18)>]> 
2.3.3 :003 > _.changes
 => {"value"=>[[#<BigDecimal:7f8c7eb21850,'0.12E1',18(27)>, #<BigDecimal:7f8c7eb21738,'0.3E1',9(18)>], [#<BigDecimal:7f8c7eb2b990,'0.12E1',18(27)>, #<BigDecimal:7f8c7eb2b710,'0.3E1',9(18)>]]} 
2.3.3 :004 > 
```

### Expected behavior
Record should not be dirty after loading from the backend

### Actual behavior
Record is `changed?` right away.

### System configuration
**Rails version**: 4.2.7

**Ruby version**: 2.3.3

**Comment by lugray** [comment]
I cannot reproduce this.  The following passes:
```ruby
# frozen_string_literal: true

begin
  require "bundler/inline"
rescue LoadError => e
  $stderr.puts "Bundler version 1.10 or later is required. Please update your Bundler"
  raise e
end

gemfile(true) do
  source "https://rubygems.org"
  gem "rails", '4.2.7'
  gem "pg"
end

require "active_record"
require "minitest/autorun"
require "logger"

# This connection will do for database-independent bug reports.
ActiveRecord::Base.establish_connection(adapter: 'postgresql', database: 'postgres-rails_test')
ActiveRecord::Base.logger = Logger.new(STDOUT)

ActiveRecord::Schema.define do
  create_table :samples, force: true do |t|
    t.string  :name,  null: true
    t.decimal :value, precision: 18, scale: 6, array: true, default: [0], null: true
  end

end

class Sample < ActiveRecord::Base
end

class BugTest < Minitest::Test
  def test_dirty
    Sample.create!(value: [1.0, 2.3])
    sample = Sample.last

    assert !sample.changed?
  end
end
```
I suspect something else in your app.  Can you provide something similar that fails?

**Comment by johvet** [comment]
Hi Lisa,

I’ve provided a minimal rails application with specs failing - so you could try that. But I wasn’t aware mini-test works with pg, so here’s what you can do to reprocude
with that:


class BugTest < Minitest::Test
  def test_dirty
    Sample.create!(value: [1.0, 2.3])
    sample = Sample.last
    sample.inspect
    assert sample.changes == {}, "changes should be empty, but are #{sample.changes.inspect}"
  end
end

I’m happy to help, so please let me know if there’s anything I can do. I think the key part is the inspect, or anything that makes the attributes actually getting pulled. It happens implicitly when in a rails console.

kind regards,

Johannes Vetter | Lead Engineer
johannes@art19.com <mailto:johannes@art19.com>

ART19, Inc. • 1299 Fourth Street Suite 505 • San Rafael CA 94901 • 415.423.3583 <tel://415.423.3583> • www.art19.com <http://www.art19.com/>
This email and any attachments thereto may contain private, confidential, and privileged material for the sole use of the intended recipient. Any review, copying, or distribution of this email (or any attachments thereto) by others is strictly prohibited. If you are not the intended recipient, please contact the sender immediately and permanently delete the original and any copies of this email and any attachments thereto.

**Comment by johvet** [comment]
As explained in https://github.com/johvet/pg_array, it seems like in `ActiveRecord::ConnectionAdapters::PostgreSQL::OID::Array#cast_value_for_database` the scale of the `BigDecimal` is lost upon turning the array into a string. When it then compares that value to the former raw_value (which hold the precision/scale enforced by the db), it must see a change.

**Comment by schallereqo** [comment]
Confirmed, this is fixed in Rails `5.1.5`
